src/agent.rs

```diff
@@ -112,6 +111,7 @@ pub(crate) struct AgentState {
     #[cfg(feature = "cookies")]
     pub(crate) cookie_tin: CookieTin,
     pub(crate) resolver: ArcResolver,
+    pub(crate) middleware: Option<Vec<Arc<dyn Middleware + Send + Sync>>>,
```

Comment:
I think we can make this just `Vec<Arc<dyn Middleware + Send + Sync>>`, since the "no middleware" case is adequately represented by an empty vec.

https://doc.rust-lang.org/std/vec/struct.Vec.html#guarantees

> if you construct a Vec with capacity 0 via Vec::new, vec![], Vec::with_capacity(0), or by calling shrink_to_fit on an empty Vec, it will not allocate memory.

Also I think we can get rid of the `Arc` if we do: `Vec<Box<dyn Middleware + Send + Sync + 'static>>`.

Reply:
Wanted to save on some unnecessary heap allocation since also an empty array would allocate capacity.

Reply:
Arc is there because we don't want to force Middleware impl to also do clone.

The concurrency in Request do_call currently builds on cloning the middleware chain, hence Arc instead of Box. 

Maybe that can change if we get the iterator in MiddleWare next working.

Reply:
> since also an empty array would allocate capacity.

But the docs I linked for Vec (https://doc.rust-lang.org/std/vec/struct.Vec.html#guarantees) guarantee that an empty array does not allocate.

Reply:
> Arc is there because we don't want to force Middleware impl to also do clone.

I think the most common case is that the Middleware impl is fully given to use - e.g. a `fn`.

There's another common case, like metrics, where the user wants an `Arc`. But in that case it's because they want to retain a handle to the middleware so they can collect data, like the example you have with Counter. But in that case, the user needs to make the Arc themselves and pass us a clone of it. So for instance in the MyCounter example we wind up with two Arcs: one built inside the Agent, and one built inside the MyCounter. It would be nicer if the user makes an `Arc<MyCounter>` and passes us a clone; that way there's only one Arc.

Maybe the best thing is if we can make a param that accepts a T or an Arc. If we get a T, we Arc::new() it; if we get an Arc, we clone it. I'm not seeing immediately what that param would look like, but I bet it exists.

Reply:
> guarantee that an empty array does not allocate.

TIL <3

Reply:
I learned it today too! My thought process was: "I bet this is to avoid allocations. But wait, does it really allocate when empty?" Very neat to learn about.

Reply:
> I think the most common case is that the Middleware impl is fully given to use - e.g. a fn.

I think I'm not making myself understood here. I will take a stab at the `iterator` approach in `MiddlewareNext` and see if we can simply side step this whole discussion :)